src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3756,7 +3757,126 @@ public RecoverWith(Func<Exception, IGraph<SourceShape<TOut>, TMat>> partialFunct
+                    if (task.IsFaulted) _onFailure(task.Exception);
+                    else if (task.IsCanceled) _onFailure(new TaskCanceledException(task));
+                    else if (task.Result) _onSuccess(enumerator.Current);
+                    else _onComplete();
+                };
+
+                SetHandler(_outlet, this);
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnComplete() => CompleteStage();
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnFailure(Exception exception) => FailStage(exception);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void OnSuccess(T element) => Push(_outlet, element);
+
+            public override void OnPull()
+            {
+                var vtask = _enumerator.MoveNextAsync();
+                if (vtask.IsCompletedSuccessfully)
+                {
+                    // When MoveNextAsync returned immediatelly, we don't need to await.
+                    // We can use fast path instead.
+                    if (vtask.Result)
+                    {
+                        // if result is true, it means we got an element. Push it downstream.
+                        Push(_outlet, _enumerator.Current);
+                    }
+                    else
+                    {
+                        // if result is false, it means enumerator was closed. Complete stage in that case.
+                        CompleteStage();
+                    }
+                }
+                else
+                {
+                    vtask.AsTask().ContinueWith(_handleContinuation);
+                }
+            }
+            public override void OnDownstreamFinish(Exception cause)
+            {
+                var vtask = _enumerator.DisposeAsync();
+                if (vtask.IsCompletedSuccessfully)
+                {
+                    CompleteStage(); // if dispose completed immediately, complete stage directly
+                }
+                else
+                {
+                    // for async disposals use async callback
+                    vtask.GetAwaiter().OnCompleted(_onComplete);
+                }
+                base.OnDownstreamFinish(cause);
+            }
+
+        }
+
+        #endregion
+        private readonly Outlet<T> _outlet = new Outlet<T>("EnumerableSource.out");
+        private readonly IAsyncEnumerable<T> _asyncEnumerable;
 
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="asyncEnumerable">TBD</param>
+        public AsyncEnumerable(IAsyncEnumerable<T> asyncEnumerable)
+        {
+            //TODO: when to dispose async enumerable? Should this be a part of ownership of current stage, or should it
+            // be a responsibility of the caller?
+            _asyncEnumerable = asyncEnumerable;
```

Comment:
@Aaronontheweb :

Not entirely related but Should we be taking a `Func<IAsyncEnumerable<T>>` here? (In my opinion yes, in ideal-state, a Source itself should be idempotent and able to be materialized multiple times.)

As to this question, for whatever my opinion is worth I'd *think* that since we are consuming the stream till the end, we should be disposing when we are done whether it is provided by a func producer or live instance.

Reply:
> Not entirely related but Should we be taking a Func<IAsyncEnumerable<T>> here? (In my opinion yes, in ideal-state, a Source itself should be idempotent and able to be materialized multiple times.)

Now that is a good question. We should absolutely program this component to accept this `Func` so we can restart the stream in the event of errors. That'll also make disposal of the `IAsyncEnumerable` less ambigious too: we created it, we own it, so we clean it up.

Reply:
@eaba can you implement this change?